File: hill_climbing.java
/*  https://www.geeksforgeeks.org/introduction-hill-climbing-artificial-intelligence/ */

// Importing libraries
import java.util.ArrayList; // เรียกใช้คลาสสำหรับสร้างจำนวนเต็ม
import java.util.Collections;
import java.util.Comparator; 
import java.util.List; // เเสดงข้อมูลที่ถูกเรียงลำดับ
import java.util.function.Function;  // นำเข้า interface ที่เเสดงถึง func ที่รับ 1 argument เเละส่งค่ากลับ

// Generates neighbors of x
public static List<Integer> generate_neighbors(int x) // function รับจำนวนเต็ม x เป็นinput เเละสร้าง (List<Integer>) เป็นจำนวนเต็มที่อยู่ไกล้เคียงกับ x
{
	// TODO: implement this function
	return new ArrayList<>();
}

// method
public static int
hill_climbing(Function<Integer, Integer> f, int x0) // ใช้รับ function f ที่รับจำนวนเต็มเป็น input เเละส่งค่าจำนวนเต็มกลับ
{
	int x = x0; // initial solution ( กำหนดค่าให้ตัวเเปรสำหรับเก็บค่า solution ปัจจุบัน เริ่มต้นด้วย x0 )
	while (true) { // Loopจนกว่าจะเจอ soluton ที่ดีที่สุด

		// find neighbors เเละ best neighbors
		List<Integer> neighbors = generate_neighbors(x); // เรียกใช้ฟังก์ชัน generate_neighbors เพื่อหา neighbors ของ x
		int best_neighbor = Collections.max(
			neighbors,
			Comparator.comparingInt(
				f::apply)); // find the neighbor with the highest function value
		// เงื่อนไขการหยุดทำงาน					
		if (f.apply(best_neighbor)
			<= f.apply(x)) // ตรวจสอบว่าค่าฟังก์ชัน f ของ best neighbor น้อยกว่าหรือเท่ากับ ค่าฟังก์ชัน f ของ x
			return x; // return x
		x = best_neighbor; // otherwise, continue with the
						// best neighbor
	}
}

public static void main(String[] args)
{
	// Example usage : ตัวอย่างการใช้งาน
	int x0 = 1; //  กำหนดค่าเริ่มต้น x0 ให้เป็น 1
	int x = hill_climbing // เรียกใช้ฟังก์ชัน hill_climbing
		((Integer y) -> y * y // ฟังก์ชัน f ที่ใช้ในอัลกอริทึม
		 , x0); 
	System.out.println("Result: " + x); // พิมพ์ค่า x
}

File: Hill_SAGroup/hill_climbing.py
import random

# ฟังก์ชันสร้าง neighbors โดยการเลื่อนค่า x ซ้ายหรือขวาอย่างสุ่ม
def generate_neighbors(x):
    return [x + random.uniform(-0.5, 0.5) for _ in range(3)] 

# ฟังก์ชัน Hill Climbing ใช้ในการค้นหาคำตอบของฟังก์ชัน f(x) ที่มีค่าน้อยที่สุด
def hill_climbing(f, x0):
    x = x0  # กำหนดค่าเริ่มต้น x
    while True:  # วน Loop จนกว่าเงื่อนไขข้างในจะเป็นเท็จ
        neighbors = generate_neighbors(x)  # สร้าง neighbors ของ x
        # หา neighbors ที่มีค่าของฟังก์ชันมากที่สุด
        best_neighbor = max(neighbors, key=f)
        # ถ้าค่าฟังก์ชันของ neighbors ที่ดีที่สุดไม่ดีกว่าหรือเท่ากับ x ให้หยุด
        if f(best_neighbor) <= f(x):
            return x
        x = best_neighbor  # มิฉะนั้น ใช้ neighbors ที่ดีที่สุดต่อไป

# ฟังก์ชัน f(x) ที่ใช้ในการคำนวนค่าของ x ยกกำลัง 2
def f(x):
    return x ** 2

# กำหนดค่าเริ่มต้น
x0 = 10  # กำหนดค่าเริ่มต้น x0 เป็น 10




# อันนี้ไม่อยู่ใน Slide ลงไปข้างล่างเลยก็ได้ #
# เรียกใช้ฟังก์ชัน hill_climbing เพื่อค้นหาคำตอบ
result = hill_climbing(f, x0)
print("Optimal value of x:", result) # พิมพ์ผลลัพธ์
print("Optimal value of f(x):", f(result))

# โค้ดด้านบนเป็นการใช้งาน Hill Climbing Algorithm เพื่อค้นหาค่า x ที่ทำให้ฟังก์ชัน f(x) = x ** 2 มีค่าน้อยที่สุด โดยเริ่มต้นที่ x0 = 10 และสร้าง neighbors
# โดยการเลื่อนค่า x ซ้ายหรือขวาอย่างสุ่ม และเลือกค่า x ที่มีค่าฟังก์ชัน f(x) มากที่สุดเป็นตัวแทนของ x ในแต่ละรอบ โดยทำซ้ำจนกว่าจะไม่มี neighbors ที่ดีกว่าค่าปัจจุบันของ x 
# หรือเงื่อนไขการหยุดเป็นเท็จ ซึ่งจะคืนค่า x ที่ได้ในที่สุดและค่า f(x) ที่ได้ในการทำงานของฟังก์ชัน hill_climbing